### PR TITLE
Define end to end scenario tests

### PR DESCRIPTION
**Pull request by peternied**
### Description
Define a specific set of scenarios for testing end to end RFS functionality.  More draft PRs for other realms of testing to arrive following this one.

### Issues Resolved
_TDB_

### Check List
- [ ] New functionality includes testing
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        //    - logs-12-3456
+        //    - data-rolling
+        //    - playground
+        //    - playground2
+        // PSEUDO: Add documents 
+        //    - 19x http-data docs into logs-01-2345 
+        //    - 23x http-data docs into logs-12-3456
+        //    - 29x data-rolling
+        //    - 5x geonames docs into playground
+        //    - 7x geopoint into playground2
```
Comment (peternied) [inline RFS/src/test/java/com/rfs/integration/EndToEndTest.java:35]: I think high-level document types are good enough for these tests, I'd rather test 'interesting' docs in lower level test cases, what do we think?

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * Tests focused on setting up whole source clusters, performing a migration, and validation on the target cluster
+ */
+public class EndToEndTest {
+
+    public Object sourceCluster;
+    public Object targetCluster;
+    public SimpleRestoreFromSnapshot simpleRfsInstance;
+
+    @Test
+    public void ES_v6_8_to_OS_v2_X_Migration() throws Exception {
```
Comment (peternied) [inline RFS/src/test/java/com/rfs/integration/EndToEndTest.java:16]: Are there other 6.8 scenarios that we'd want end to end coverage on or is 1 enough?

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        // PSEUDO: Verfiy global cluster state
+        //    - 6.8 Setting is transformed
+        // PSEUDO: Verify creation of 2 index templates on the cluster
+        // PSEUDO: Verify creation of 5 indices on the cluster
+        //    - logs-01-2345
+        //    - logs-12-3456
+        //    - data-rolling
+        //    - playground
+        //    - playground2
+        // PSEUDO: Verify documents
```
Comment (peternied) [inline RFS/src/test/java/com/rfs/integration/EndToEndTest.java:55]: Yes this is hand-wavy, I think we might do this roughly to start and then refactor/improve if we find it useful

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        
+        // PSEUDO: Additional validation:
+        //   - Mapping type parameter is removed
+        // 
+
+        // Looking for inspriation?
+        //   - OS Breaking Changes https://opensearch.org/docs/latest/breaking-changes/
+        //   - ES Breaking Changes https://www.elastic.co/guide/en/elasticsearch/reference/7.0/breaking-changes-7.0.html
```
Comment (peternied) [inline RFS/src/test/java/com/rfs/integration/EndToEndTest.java:63]: Getting granular on these items will be something that should be isolated more towards the transformation test cases, but still feel that maybe there is something to draw from and include here.

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java (hunk 3 above)
Comment (chelma) [reply RFS/src/test/java/com/rfs/integration/EndToEndTest.java:55]: That's fine w/ me.  Even something as simple as:
* The doc exists w/ the same ID
* The doc matches what we'd expect after performing a transform on the original doc (e.g. nothing changed its contents besides the transformer)

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Test
+    public void ES_v6_8_to_OS_v2_X_Migration() throws Exception {
+        // Setup
+        // PSEUDO: Create a source cluster running ES 6.8
+        // PSEUDO: Update global cluster state
+        //    - Change a 6.8 relevant setting
+        // PSEUDO: Create 2 index templates on the cluster
+        //    - logs-*
+        //    - data-rolling
```
[the remark this answers is not in the document]
Comment (peternied) [reply RFS/src/test/java/com/rfs/integration/EndToEndTest.java:23]: Restating in my own works, It sounds like there is a concept 'index templates' that has a specific implementation in ES7 vs ES6.  Do you have some reference on how these are declared in ES6?

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    public Object sourceCluster;
+    public Object targetCluster;
+    public SimpleRestoreFromSnapshot simpleRfsInstance;
+
+    @Test
+    public void ES_v6_8_to_OS_v2_X_Migration() throws Exception {
+        // Setup
+        // PSEUDO: Create a source cluster running ES 6.8
+        // PSEUDO: Update global cluster state
+        //    - Change a 6.8 relevant setting
```
[the remark this answers is not in the document]
Comment (peternied) [reply RFS/src/test/java/com/rfs/integration/EndToEndTest.java:20]: Maybe I should pull this section until we have something more concrete defined, are there any cluster settings on the source cluster that should be taken into consideration when configuring the target cluster?

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        // PSEUDO: Create 2 index templates on the cluster
+        //    - logs-*
+        //    - data-rolling
+        // PSEUDO: Create 5 indices on the cluster
+        //    - logs-01-2345
+        //    - logs-12-3456
+        //    - data-rolling
+        //    - playground
+        //    - playground2
+        // PSEUDO: Add documents
```
[the remark this answers is not in the document]
Comment (peternied) [reply RFS/src/test/java/com/rfs/integration/EndToEndTest.java:30]: Good idea, we can add them to the 'set of documents that need to be migrated'.  Note; there should be much more exclusive and individualized testing for scenarios such as the relationship between the parent and child doc ids aligning that seem better suited to transformation specific tests.

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        //    - playground
+        //    - playground2
+        // PSEUDO: Add documents 
+        //    - 19x http-data docs into logs-01-2345 
+        //    - 23x http-data docs into logs-12-3456
+        //    - 29x data-rolling
+        //    - 5x geonames docs into playground
+        //    - 7x geopoint into playground2
+
+        // PSEUDO: Create a target cluster running OS 2.X (Where x is the latest released version)
```
[the remark this answers is not in the document]
Comment (peternied) [reply RFS/src/test/java/com/rfs/integration/EndToEndTest.java:37]: You raise a good point, I think this also hits on the conversation we had about 'snapping to versions'. However, I think we should treat the migration target differently than the sources.

- Sources should be very distinct, maybe even have a full set of minor versions level tests.

- Targets would follow the OpenSearch support model is specifically for the N-1 major, most recent minor and patch (v1.3.16) and N major, most recent minor and patch (v2.14.0).

I think we will have a large set of source versions that are specifically supported and hopefully a very small number of target versions (why wouldn't folks want to migrate to a supported version?)

@sumobrian Do you know if we have a position on this?

[the remark this answers is not in the document]
Comment (peternied) [reply RFS/src/test/java/com/rfs/integration/EndToEndTest.java:37]: > [Recap offline discussion] Per @sumobrian we should support multiple target cluster versions - for the moment they are the latest OS 1.x and 2.x..

I'll update the test cases to cover these scenarios

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (chelma) [reply RFS/src/test/java/com/rfs/integration/EndToEndTest.java:20]: That's an excellent question.  So far we've assumed that settings our outside the domain of the migration, but that's just an assumption and may change with time.  I'd say take it out for now; I'm confident it will pop back up soon if it's something we actually need to tackle.

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (chelma) [reply RFS/src/test/java/com/rfs/integration/EndToEndTest.java:23]: Yeah, they're just called "templates".  Starting in ES 7.8, they are split into "index" and "component" templates, where index templates can be composed of component templates.  I'd just check the Elastic docs for 6.8 vs. 7.10 or something.

**Review by chelma — COMMENTED** [review]
Thanks for doing this, Peter!

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        // PSEUDO: Verify creation of 5 indices on the cluster
+        //    - logs-01-2345
+        //    - logs-12-3456
+        //    - data-rolling
+        //    - playground
+        //    - playground2
+        // PSEUDO: Verify documents
+        
+        // PSEUDO: Additional validation:
+        //   - Mapping type parameter is removed
```
[the remark this answers is not in the document]
Comment (peternied) [reply RFS/src/test/java/com/rfs/integration/EndToEndTest.java]: In 7.X [1] there was a feature called types that was deprecated.  So while allowed in 6.8 - it was scheduled for removal in 7.X, and its gone in ES 8.X and OS 2.X 

The logic I am using is to working backwards from when using 6.8 mapping types were allowed so we should make sure the transfer happens well with the migration tooling.  There is nothing preventing migration of 7.X -> `{new version}` from also doing the same transformation, but its less supported / maybe we could worry less about testing it?

- [1] https://www.elastic.co/guide/en/elasticsearch/reference/7.17/removal-of-types.html

[the remark this answers is not in the document]
Comment (peternied) [reply RFS/src/test/java/com/rfs/integration/EndToEndTest.java]: This list [1] is a great starting point for the 'interesting' scenarios when targeting OpenSearch versions  

- [1] https://opensearch.org/docs/latest/breaking-changes/

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+    protected Object sourceCluster;
+    protected Object targetCluster;
+    protected SimpleRestoreFromSnapshot simpleRfsInstance;
+
+    @ParameterizedTest(name = "Target OpenSearch {0}")
+    @ArgumentsSource(SupportedTargetCluster.class)
+    @Disabled
+    public void migrateFrom_ES_v6_8(final OpenSearchContainer.Version targetVersion) throws Exception {
+        // Setup
```
Comment (gregschohn) [inline RFS/src/test/java/com/rfs/integration/EndToEndTest.java:28]: one last comment on building this out.  It would be great if the way to generate this were all through REST commands so that we could spin it around to use for CDC tests too.

File: RFS/src/test/java/com/rfs/framework/OpenSearchContainer.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+    @SuppressWarnings("resource")
+    public OpenSearchContainer(final Version version) {
+        this.version = version;
+        container = new GenericContainer<>(DockerImageName.parse(this.version.imageName))
+                .withExposedPorts(9200)
+                .withEnv("discovery.type", "single-node")
+                .withEnv("plugins.security.disabled", "true")
+                .withEnv("OPENSEARCH_INITIAL_ADMIN_PASSWORD", "SecurityIsDisabled123$%^")
+                .waitingFor(Wait.forHttp("/").forPort(9200).forStatusCode(200).withStartupTimeout(Duration.ofMinutes(1)));
```
[the remark this answers is not in the document]
Comment (peternied) [reply RFS/src/test/java/com/rfs/framework/OpenSearchContainer.java:28]: 9200 is the port inside the container, its dynamically mapped outside which can be picked up by using `container.getMappedPort(9200);` as is done in `getUrl()`

File: RFS/src/test/java/com/rfs/framework/OpenSearchContainer.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    public void close() throws Exception {
+        log.info("Stopping version: " + version);
+        log.debug("Instance logs:\n" + container.getLogs());
+        container.stop();
+    }
+
+    public static enum Version {
```
[the remark this answers is not in the document]
Comment (peternied) [reply RFS/src/test/java/com/rfs/framework/OpenSearchContainer.java:49]: This was intentional, if the code needs different versions numbers we should probably have a good reason and I'd prefer a compile time safe check that everything is in alignment.  For example we could shift the from `2.14.0` -> `2.15.0` and we would known that those `2.14.0` references aren't in use any more - a singular authority.

I'm sure we can examine this over time.  But I am a big fan of there being a finite number of places where we can pick a version.

Note; maybe we can iterate in future PRs to rename these to `V1_3_CURRENT` for the enum name, but the version number it maps to is still specific if that makes more sense

File: RFS/src/test/java/com/rfs/framework/SimpleRestoreFromSnapshot.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+    public static SimpleRestoreFromSnapshot forCluster(final String sourceClusterUrl) {
+        // TODO: determine version from source cluster
+        return new SimpleRestoreFromSnapshot_ES_7_10();
+    }
+
+    public default void fullMigrationViaLocalSnapshot(final String targetClusterUrl) throws Exception {
+        // TODO: Dynamically create / clean these up during tests
+        final var tempSnapshotName = "";
+        final var compressedSnapshotDirectory = "";
```
[the remark this answers is not in the document]
Comment (peternied) [reply RFS/src/test/java/com/rfs/framework/SimpleRestoreFromSnapshot.java:20]: I am hearing an undercurrent of an assumption that the end to end tests **need** to be fast - we can't make them fast (<1 sec each) if they are using a full docker instance of the source and target cluster. 

If we see a good opportunity to increase our velocity by speeding up these tests that is a good parallelable effort that we can circle back on.

---

PS I think we should be really careful about how many test cases we will have and make sure they are the 'right ones' at this level.  Its important that test validation time doesn't get longer without a clear tradeoff.

File: RFS/src/test/java/com/rfs/integration/EndToEndTest.java (hunk 10 above)
Comment (peternied) [reply RFS/src/test/java/com/rfs/integration/EndToEndTest.java:28]: I think we need to test different things in different systems - happy path we could have good alignment, but edge cases for CDC will need much more flexiblity.  That said - I'll keep in mind good opportunities for reuse, thanks for bring this up.
